### ISSUE TITLE
[Bug] podAffinity for Galera recovery jobs

### ISSUE THREAD

**Issue by 0xmeyer** [mariadb-operator/mariadb-operator#840]
**Documentation**
- [x] I acknowledge that I have read the relevant [documentation](https://github.com/mariadb-operator/mariadb-operator/tree/main/docs).

**Describe the bug**
I'm currently testing the mariadb-operator to run a Galera cluster in Kubernetes and I'm currently trying out the recovery process. Unfortunately, this does not work as described.

I have a StorageClass that only supports RWO. As soon as I kill all Galera pods, the recovery process is started after approx. 30s and therefore a job is deployed for each instance. However, since it is currently not possible to configure `podAffinity` for those jobs, it is possible that the jobs are scheduled on a different node than the PVC is currently mounted, which means that the recovery process does not complete. It would therefore be helpful if it would be possible to configure `podAffinity` for the job or enable it by default.

**Comment by mmontes11** [comment]
Hey there @0xmeyer !  Thanks for bringing this up! Makes total sense to me.

I suggest we introduce an `affinity` field in the recovery `Job`:

```yaml
apiVersion: k8s.mariadb.com/v1alpha1
kind: MariaDB
metadata:
  name: mariadb-galera
spec:
  galera:
    recovery:
       job:
         affinity:
           podAffinityEnabled: true  
```
`podAffinityEnabled` will be defaulted to true if not provided by the user, and it will result in the following affinity rules:

```yaml
apiVersion: k8s.mariadb.com/v1alpha1
kind: MariaDB
metadata:
  name: mariadb-galera
spec:
  galera:
    recovery:
      job:
        affinity:
          podAffinityEnabled: true
          podAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
              - labelSelector:
                  matchExpressions:
                  - key: app.kubernetes.io/instance
                    operator: In
                    values:
                    - mariadb-galera
                topologyKey: kubernetes.io/hostname
```
This way, we are enforcing the `Job` to be scheduled in the same node as the `MariaDB` `Pod`. 

Do you think this would cover your case?

**Comment by 0xmeyer** [comment]
> I suggest we introduce an `affinity` field in the recovery `Job`

Sounds good to me.

> and it will result in the following affinity rules ... we are enforcing the Job to be scheduled in the same node as the MariaDB Pod

Since every pod in the Galera cluster has this label, we can't make sure that the `mariadb-galera-0-recovery` pod is scheduled on the same node as the `mariadb-galera-0` pod, can we? Maybe we should additionally use the labels created by the statefulset controller (e.g. `apps.kubernetes.io/pod-index`). However, `apps.kubernetes.io/pod-index` cannot be defined statically; the operator must have logic built into it.

```yaml
apiVersion: k8s.mariadb.com/v1alpha1
kind: MariaDB
metadata:
  name: mariadb-galera
spec:
  galera:
    recovery:
      job:
        affinity:
          podAffinityEnabled: true
          podAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
              - labelSelector:
                  matchExpressions:
                  - key: app.kubernetes.io/instance
                    operator: In
                    values:
                    - mariadb-galera
                  - key: apps.kubernetes.io/pod-index
                    operator: In
                    values:
                    - "0"
                topologyKey: kubernetes.io/hostname
```

**Comment by mmontes11** [comment]
> Since every pod in the Galera cluster has this label, we can't make sure that the `mariadb-galera-0-recovery` pod is scheduled on the same node as the `mariadb-galera-0` pod, can we? Maybe we should additionally use the labels created by the statefulset controller (e.g. `apps.kubernetes.io/pod-index`). However, `apps.kubernetes.io/pod-index` cannot be defined statically; the operator must have logic built into it.

God point, but I think this should be covered by the `topologyKey=kubernetes.io/hostname` right? The hostname will already contain the index, for example: `mariadb-galera-0`, and therefore there is no need to match it with another rule.

**Comment by mmontes11** [comment]
Actually... the hostname for the recovery `Pods`  would be `mariadb-galera-0-recovery` which doesn't match the `MariaDB` `Pods`:  `mariadb-galera-0`. 

> However, apps.kubernetes.io/pod-index cannot be defined statically; the operator must have logic built into it.

Anyway, I think I need to test this out, but I got the idea. Adding an extra expression based on the `Pod` index from the operator side would not be an issue, if really needed.

**Comment by mmontes11** [comment]
@0xmeyer that's correct:

```go
type PodAffinityTerm struct {
	// A label query over a set of resources, in this case pods.
	// If it's null, this PodAffinityTerm matches with no Pods.
	// +optional
	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty" protobuf:"bytes,1,opt,name=labelSelector"`
	// This pod should be co-located (affinity) or not co-located (anti-affinity) with the pods matching
	// the labelSelector in the specified namespaces, where co-located is defined as running on a node
	// whose value of the label with key topologyKey matches that of any node on which any of the
	// selected pods is running.
	// Empty topologyKey is not allowed.
	TopologyKey string `json:"topologyKey" protobuf:"bytes,3,opt,name=topologyKey"`
	...
}
``` 
TL;DR; we need to match `apps.kubernetes.io/pod-index` to ensure it runs on the same node, as you pointed out [here](https://github.com/mariadb-operator/mariadb-operator/issues/840#issuecomment-2337879288).

**Comment by mmontes11** [comment]
@0xmeyer this has been supported via https://github.com/mariadb-operator/mariadb-operator/pull/843. Stay tuned for the next release!

**Comment by 0xmeyer** [comment]
@mmontes11 awesome! Thank you!

**Comment by mmontes11** [comment]
Out of curiosity @0xmeyer, which storage provider you use?

**Comment by mmontes11** [comment]
Thanks for confirming, got another user using Longhorn with the same problem: https://github.com/mariadb-operator/mariadb-operator/issues/848

Any RWO will have this problem, just wanted to rule out that was something Longhorn specific

**Comment by mmontes11** [comment]
[v0.0.31](https://github.com/mariadb-operator/mariadb-operator/releases/tag/v0.0.31) has been released with this fix. Please follow [this guide ](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/UPGRADE_v0.0.31.md)to upgrade.

**Comment by 0xmeyer** [comment]
Hi @mmontes11, it works perfectly. Many thanks!

**Comment by mmontes11** [comment]
@0xmeyer thanks for the detailed report and helping with the troubleshooting! Much appreciated 🙏🏻
